### DISCUSSION TITLE
Can long living effects be cancelled automatically when a view disappears?

### DISCUSSION THREAD

**Discussion by vctxr** [pointfreeco/swift-composable-architecture#1415] [Q&A] [answered]
Hi thank you so much for creating this amazing library.

I have a view that executes a long living effect (a timer). but when i tap on back to dismiss the view, the timer still executes even when the view is no longer visible. After checking it, the store is not deallocated when i tap on the back button causing the timer to still run.

Are there a way so i can automatically cancel all my long living effects when my view disappears?

Thank you, any help and guidance would be very helpful. Here is my sample code:
```swift
@main
struct DemoApp: App {
    var body: some Scene {
        WindowGroup {
            View1()
        }
    }
}

struct View1: View {
    var body: some View {
        NavigationView {
            List {
                NavigationLink("Navigate to View2", destination: View2())
            }
            .navigationTitle("View1")
        }
    }
}

struct View2: View {
    private let store: Store<Void, Action> = Store(initialState: (), reducer: reducer, environment: ())
    
    var body: some View {
        Text("View2")
            .onAppear {
                ViewStore(store).send(.onAppear)
            }
    }
}

enum Action {
    case onAppear
    case receiveLongLivingEffect
}

let reducer = Reducer<Void, Action, Void> { _, action, _ in
    switch action {
    case .onAppear:
        return Effect.timer(id: "id", every: 1, on: DispatchQueue.main)
            .map { _ in Action.receiveLongLivingEffect }
            .eraseToEffect()
        
    case .receiveLongLivingEffect:
        print(">> Receive effect")
        return .none
    }
}
```

**Comment by mbrandonw** [comment, marked as answer]
Hi @vctxr, if you can use version [0.39.0](https://github.com/pointfreeco/swift-composable-architecture/releases/tag/0.39.0) or later of the library, then this can automatically be handled for you via Swift's concurrency tools.

Rather than using `.onAppear` you need to use SwiftUI's `.task` view modifier, which allows you to perform some asynchronous work when the view appears, and that work will automatically be cancelled when the view goes away.

So, rather than having an `onAppear` action that uses a Combine timer, you can have a `.task` action that makes use of an async/await timer:

```swift
case .task: 
  return .run { send in 
    for await _ in DispatchQueue.main.timer(interval: .seconds(1)) {
      await send(.receiveLongLivingEffect)
    }
  }
```

And then you can `await` sending that action in the view:

```swift
.task {
  await viewStore.send(.task).finish()
}
```

Then when the view goes away this timer will automatically be cancelled and the effect torn down.

I would also recommend thinking about putting a scheduler in your environment to use for the timer rather than reaching out to `DispatchQueue.main` directly. That will make it possible to write tests for your feature later.

**Reply by vctxr** [reply]
thanks for the detailed explanation @mbrandonw . i see that the `.task` modifier of `SwiftUI` automatically cancels the effect when the view goes away. But what if the effect is triggered not from a `.task` modifier, but let's say from a button tap:
```swift
Button("Start timer") {
    viewStore.send(.onButtonTap)
}
```

 In that case is there a way for the effect to automatically be cancelled too? also, thank you for the suggestion to put a scheduler in the environment 👍

**Reply by rcarver** [reply]
Hi @vctxr I shared a method of cancelling other tasks [in this thread](https://github.com/pointfreeco/swift-composable-architecture/discussions/1327#discussioncomment-3531685)

**Reply by mbrandonw** [reply]
Yeah, in order to cancel all effects that were started from _all_ actions, you need to do more work. If your feature isn't being driven off of optional state, then you can tap into `onDisappear` to cancel all effects.

But our navigation APIs are really the thing that is going to solve this once-and-for-all, but that's still an active research topic and needs more time.

**Reply by vctxr** [reply]
thanks for the insightful answers @rcarver and @mbrandonw. For now, that method does the trick of cancelling effects, and i'm also looking forward to what TCA's navigation API will achieve.

**Reply by jslew** [reply]
@mbrandonw Is there something you can point to now that explains the current situation since navigation has been released and we are near tca 1.0?  Thanks!

**Reply by mbrandonw** [reply]
The bottom of the docs for [`ifLet`](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/scope/iflet(_:action:destination:fileid:line:)) describe all the things that operator enforces, including automatic effect cancellation when the state is `nil`'d out.

**Reply by stephencelis** [reply]
@jslew All optional/enum/collection-based reducer APIs (`.ifLet`, `.forEach`, `.ifCaseLet`) automatically tear down their effects these days. Each API should document this, but if we're missing something, let us know!

**Reply by jslew** [reply]
Thanks, good info.  The individual API documentation is outstanding, but it can be difficult to be able to get to the right piece from a conceptual point-of-view.   Perhaps this is already called out in the top-level documentation topics, but maybe some discussion of this in the "Concurrency" section of the main documentation would be a good addition.
